core/src/main/java/com/linecorp/armeria/common/stream/AsyncMapStreamMessage.java

```diff
@@ -240,13 +247,14 @@ public void request(long n) {
         private void handleRequest(long n) {
             requestedByDownstream = LongMath.saturatedAdd(requestedByDownstream, n);
 
-            if (state == State.WAITING) {
+            if (maxConcurrency > LongMath.saturatedAdd(requestedFromUpstream, pendingFutures)) {
+                final long toRequest = maxConcurrency - pendingFutures - requestedFromUpstream;
                 if (requestedByDownstream != Long.MAX_VALUE) {
-                    requestedByDownstream--;
+                    requestedByDownstream -= toRequest;
```

Comment:
Let's say we have the following condition:
```
n = 1
maxConcurrency = 10
requestedByDownstream, requestedFromUpstream, pendingFutures = 0
```
If so, `toRequest` becomes `10` and `requestedByDownstream` becomes `-10`.
Consequently, `upstream.request(10)` is called even if a `Subscriber` requests 1 element.

Reply:
Ah, somehow I forgot about the possibility that `n < maxConcurrency`